src/domain/tradingview/TVRequests.ts

```diff
@@ -0,0 +1,100 @@
+import { LAST_BAR_REFRESH_INTERVAL, SUPPORTED_RESOLUTIONS } from "config/tradingview";
+import { getLimitChartPricesFromStats, getStablePriceData, timezoneOffset } from "domain/prices";
+import { CHART_PERIODS, USD_DECIMALS } from "lib/legacy";
+import { formatAmount } from "lib/numbers";
+import { getCurrentPriceOfToken, getTokenChartPrice } from "./requests";
+import { formatTimeInBar } from "./utils";
+
+type Bar = {
+  time: number;
+  open: number;
+  high: number;
+  low: number;
+  close: number;
+  ticker?: string;
+  volume?: number;
+};
+
+export class TVRequests {
+  lastBar: Bar | null;
+  startTime: number;
+  lastTicker: string;
+  lastPeriod: string;
+
+  constructor() {
+    this.lastBar = null;
+    this.startTime = 0;
+    this.lastTicker = "";
+    this.lastPeriod = "";
+  }
+
+  async getLastBar(chainId, ticker, period) {
+    if (!ticker || !period || !chainId) {
+      throw new Error("Invalid input. Ticker, period, and chainId are required parameters.");
+    }
+    const currentTime = Date.now();
+    if (
+      currentTime - this.startTime > LAST_BAR_REFRESH_INTERVAL ||
+      this.lastTicker !== ticker ||
+      this.lastPeriod !== period
+    ) {
+      const prices = await getLimitChartPricesFromStats(chainId, ticker, period, 1);
+      if (prices?.length) {
+        this.lastBar = formatTimeInBar({ ...prices[prices.length - 1], ticker });
+        this.startTime = currentTime;
+        this.lastTicker = ticker;
+        this.lastPeriod = period;
+      }
+    }
+    return this.lastBar;
+  }
+
+  async getHistoryBars(chainId: number, ticker: string, resolution: string, isStable: boolean, countBack: number) {
+    const period = SUPPORTED_RESOLUTIONS[resolution];
+    try {
+      const bars = isStable ? getStablePriceData(period, countBack) : await getTokenChartPrice(chainId, ticker, period);
+      return bars.map(formatTimeInBar);
+    } catch {
+      throw new Error("Failed to get history bars");
+    }
+  }
+
+  async getLiveBar(chainId: number, ticker: string, resolution: string) {
+    const period = SUPPORTED_RESOLUTIONS[resolution];
+    if (!ticker || !period || !chainId) return;
+    const periodSeconds = CHART_PERIODS[period];
+    // Converts current time to seconds, rounds down to nearest period, adds timezone offset, and converts back to milliseconds
+    const currentCandleTime = (Math.floor(Date.now() / 1000 / periodSeconds) * periodSeconds + timezoneOffset) * 1000;
+    try {
+      this.lastBar = await this.getLastBar(chainId, ticker, period);
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error(error);
+    }
+
+    if (!this.lastBar) return;
+
+    const currentPrice = await getCurrentPriceOfToken(chainId, ticker);
+    const averagePriceValue = parseFloat(formatAmount(currentPrice, USD_DECIMALS, 4));
+    if (this.lastBar.time && currentCandleTime === this.lastBar.time && ticker === this.lastBar.ticker) {
+      return {
```

Comment:
just to confirm, in this case you don't need to update `this.lastBar`, right?

Reply:
Yes, we dont need to update the lastBar in this case